Add '-r' option to Parser so directories can be searched for tests recursively

davidcpell opened the pull request:
Hi! First time attempting to contribute something besides readme changes/fixes to an open source project so bear with me. `m` has made my life with Minitest a lot easier but from time to time I've wanted to run a "chunk" of the test suite (like all of the model tests) but passing a directory to `m` just looks for all of the `*test.rb` files in the directory. I added this `-r` flag that gets all of the test files in the passed directory and all of its subdirectories, allowing me to do `m -r test/models/`and get every test in every subdirectory of `models/`. 

If this feature is of any interest to the maintainers, I was wondering if someone might be able to take a look at the test I've written and make sure it's doing what I intended? I'm having a hard time wrapping my head around the test setup (never used `appraisals`, think I understand how the tests work with different frameworks but not sure). 

The number `15` that I hardcoded into the test passes, but I want to make sure I'm counting correctly based on which tests get run and which don't depending on the Framework being used (Minitest 4/5, TestUnit). I tried to write some more specific tests where I was sure about the number of tests that would be called by `m` but ran into various problems with things not being loaded correctly and debugging tools that I'm used to using in these situations not working as expected. I've tested this is an application I'm building at work and on a dummy set of tests and it seems to be working fine there.

zamith commented on lib/m/parser.rb:
```diff
@@ -67,8 +67,17 @@ def parse_options!(argv)
     end
+
+    def wildcard(type)
+      "#{testable.file}/" + (testable.recursive? ? "**/*#{type}*.rb" : "*#{type}*.rb")
```
What about having this as an explicit if statement, even if that means having to repeat the `testable.file` part?

zamith commented on lib/m/testable.rb:
```diff
@@ -1,15 +1,20 @@
     end
+
+    def recursive?
+      recursive == true
```
This can just be `recursive`, right?

zamith commented:
@davidcpell First of all thanks a lot for contributing and I hope you contribute to many other projects. Cheers for that.

Testing with all these different setup can get tricky, it's definitely not ideal, but we have to able to test all of them. That being said, you mention some other tests you wrote, but weren't able to make pass, can you share those as well? We can try to make it work together.

I probably won't be very responsive the next few days, so happy New Year and talk to you later.

davidcpell commented:
@zamith thanks so much for your reply. I'm going to refactor according to your comments, give the tests another shot, and get back to you next week. Happy New Year!

davidcpell commented:
@zamith I've spent some more time with it and think I've made progress. 

I realized I was using my system `m` to run my individual test rather than `bin/m`, which was leading to some issues. Also, I kept getting unexpected changes in the `gemfiles/*.gemfile.lock` files when running `bundle exec rake tests` related to the `m` version. I had been just reverting those changes before committing but this time I kept them in a separate commit and thought I'd just ask you about it. :)

I removed `Testable#recursive?` and refactored `Parser#wildcard` based on your suggestions. I also added a bit to `README.md`. The most important difference, of course, is that I added a few new subdirectories under `test/examples/subdir` with new dummy test files being used by `EverythingTest#test_running_tests_recursively` to show that the flag works as intended. 

When making my dummy tests I just used the `MTest` class for my test class inheritance as I saw in some of the files, but I didn't ever use `ActiveSupport::TestCase` or `Test::Unit::TestCase`. Is that something that would need to be changed?

zamith commented:
@davidcpell I'm guessing the `gemfile.lock` issues are due to having bumped the version and not running all tests locally, or something like that. But I see no problem in merging those in.

As for `MTest`, you can see [here](https://github.com/qrush/m/blob/master/test/test_helper.rb#L35) that it will be an alias for the appropriate version of minitest or test unit, so you can just use that. I don't see how this feature could affect `AS`, so no need to add extra examples for it.

I'll take a look at the actual code when I have some time (probably Friday). One thing that I'm interested in, is the performance hit we take with this, if any. We'll have to test it against the benchmarks we already have.

In the meantime, if you can try to figure out why the tests are failing for ruby `2.2.0`, that would be great.

davidcpell commented:
@zamith I'm still working on trying to figure out what's going on with the `2.2.0`, will try to get back to you on that ASAP

zamith commented:
@davidcpell Can you rebase with master and try this again, please?

davidcpell commented:
@zamith sure thing. Also just wanted to say I noticed I should've done this on a feature branch in my fork...sorry about that

zamith commented on lib/m/parser.rb:
```diff
@@ -67,8 +67,21 @@ def parse_options!(argv)
         end
 
+        opts.on '-r', 'Search provided directory recursively.' do |line|
+          p "Searching provided directory for tests recursively"
```
I don't think we want to print this message. Do you felt a need for it?

davidcpell replied:
No, not really. I can mark that to be removed

davidcpell commented:
That seems to have done the trick for the 2.2.0 build! 

One thing that has been on my mind is that an exception will be raised if a user tries to use `m -r` by itself with no other args. This is because the `argv.empty?` returns `false` in [line 12 of Parser](/lib/m/parser.rb#L12) due to `-r`, but then in line 19 `argv.first` is `nil`. I'm not sure of the cleanest way to avoid this. So far I've thought of adding some kind of guard clause to warn the user if `-r` is passed with no other args, or changing the logic in line 12 to say `if argv.empty || argv == ["-r"]`. 

This also leads to question of whether/how to hand cases where a user might for some reason use `-r` with another option,  `-l`.  Is this overthinking it?

zamith commented:
@davidcpell I have run the benchmarks and I've seen no noticeable changes, which is nice.

zamith commented on lib/m/parser.rb:
```diff
@@ -67,8 +67,21 @@ def parse_options!(argv)
           testable.line = line
         end
 
+        opts.on '-r', 'Search provided directory recursively.' do |line|
```
I think we can change this to support the option we need it to:

```ruby
opts.on '-r', '--recursive DIR', 'Search provided directory recursively.' do |directory|
  testable.recursive = true
  argv << directory
end
```

This will throw `OptionParser::MissingArgument`, as expected, when running `m -r` without any argument.

What do you think?

davidcpell commented:
@zamith Thanks for the feedback. That drove me to look more closely at the `OptionParser` API and I have a better understanding now of how it works. I made the change and added a second test to verify. Also, I moved the two tests to `test/options_test.rb`

davidcpell commented:
That was my initial thought, too. But when I used `assert_raises` the test fails, saying no exceptions were raised. I think this is because `OptionParser` is rescuing and printing its failure messages to stdout. Not familiar with the internals of the library but I did very with a quick check that rescuing an exception causes `assert_raises` to fail:
```
  def test_rescuing_exceptions_circumvents_assert_raises
    assert_raises RuntimeError do 
      begin
        fail
      rescue => e
        puts "rescued!"
      end
    end
  end
```
```
Finished tests in 2.391279s, 3.3455 tests/s, 9.2001 assertions/s.

  1) Failure:
OptionsTest#test_rescuing_exceptions_circumvents_assert_raises [/Users/davidpell/code/ruby/metal/test/options_test.rb:42]:
RuntimeError expected but nothing was raised.
```

zamith commented:
Right, that makes sense. Let's go with the output assertion then.

zamith commented on test/options_test.rb:
```diff
@@ -30,4 +30,14 @@ def test_line_option_has_precedence_over_colon_format
+
+  def test_recursive_option_without_directory_arg_raises_exception
+    output = m('-r')
+    assert_match(/OptionParser::MissingArgument/, output)
```
Actually, can this be an `assert_output` then? To match the other tests.

davidcpell commented:
I had tried that also but the test would have fail the `$?.success?` check because the exit status of the test process was `1`.

I've gone back in this most recent iteration though added an `#assert_failure_output` helper that does the same thing but expects the process not to have exited successfully.  Thoughts?

zamith commented:
@davidcpell Oh, I thought this was using Minitest's own `assert_output`, not a custom one, sorry about that. So, that's what we should be using I think.

But that's not in the scope of this PR, so I'd say rollback the last commit (sorry :/) and I'll merge this.

davidcpell commented:
haha no worries! I'd be happy to help with that stuff, too

zamith commented:
Merged in 80f1517. Thanks for the PR! :+1:
